// stylelint-react-native/dist/rules/css-property-no-unknown/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = _default;
exports.messages = exports.ruleName = void 0;

var _reactNativeKnownStylingProperties = require("react-native-known-styling-properties");

var _stylelint = require("stylelint");

var _utils = require("../../utils");

var ruleName = (0, _utils.namespace)("css-property-no-unknown");
exports.ruleName = ruleName;

var messages = _stylelint.utils.ruleMessages(ruleName, {
  rejected: function rejected(property) {
    return "Unexpected unknown property \"".concat(property, "\"");
  }
});

exports.messages = messages;

var props = _reactNativeKnownStylingProperties.allCSS2RNProps.map(_utils.kebabCase);

function _default(actual, options) {
  return function (root, result) {
    var validOptions = _stylelint.utils.validateOptions(result, ruleName, {
      actual: actual
    }, {
      actual: options,
      possible: {
        ignoreProperties: [_utils.isString]
      },
      optional: true
    });

    if (!validOptions) {
      return;
    }

    root.walkDecls(function (decl) {
      var prop = decl.prop;

      if (!(0, _utils.isStandardSyntaxProperty)(prop)) {
        return;
      }

      if (!(0, _utils.isStandardSyntaxDeclaration)(decl)) {
        return;
      }

      if ((0, _utils.isCustomProperty)(prop)) {
        return;
      }

      if ((0, _utils.isExportBlock)(decl.parent)) {
        return;
      }

      if ((0, _utils.optionsMatches)(options, "ignoreProperties", prop)) {
        return;
      }

      if (props.indexOf(prop.toLowerCase()) !== -1) {
        return;
      }

      if (/^-styled-mixin\d+/.test(prop.toLowerCase())) {
        return;
      }

      _stylelint.utils.report({
        message: messages.rejected(prop),
        node: decl,
        result: result,
        ruleName: ruleName
      });
    });
  };
}
