// stylelint-react-native/dist/rules/font-weight-no-ignored-values/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = _default;
exports.messages = exports.ruleName = void 0;

var _stylelint = require("stylelint");

var _utils = require("../../utils");

var declarationValueIndex = require("stylelint/lib/utils/declarationValueIndex");

var ruleName = (0, _utils.namespace)("font-weight-no-ignored-values");
exports.ruleName = ruleName;

var messages = _stylelint.utils.ruleMessages(ruleName, {
  rejected: function rejected(weight) {
    return "Unexpected font-weight \"".concat(weight, "\"");
  }
});

exports.messages = messages;
var acceptedWeights = ["400", "700", "normal", "bold"];

function _default(actual) {
  return function (root, result) {
    var validOptions = _stylelint.utils.validateOptions(result, ruleName, {
      actual: actual
    });

    if (!validOptions) {
      return;
    }

    root.walkDecls(/^font-weight$/i, function (decl) {
      if (acceptedWeights.indexOf(decl.value) > -1) {
        return;
      }

      var weightValueOffset = decl.value.indexOf(decl.value);
      var index = declarationValueIndex(decl) + weightValueOffset;

      _stylelint.utils.report({
        message: messages.rejected(decl.value),
        node: decl,
        result: result,
        ruleName: ruleName,
        index: index
      });
    });
  };
}
